### PR TITLE
Annotating libraries with small number of warnings to be AOT safe

### PR DESCRIPTION
**Pull request by LakshanF**
Draft PR to check the viability to make these libraries AOT-compatible with minimal effort

- System.Reflection.DispatchProxy: A call to [AssemblyBuilder.DefineDynamicAssembly](https://github.com/dotnet/runtime/blob/main/src/libraries/System.Reflection.DispatchProxy/src/System/Reflection/DispatchProxyGenerator.cs#L130) is marked with RequiresDynamicCodeAttribute and propagated
- System.ComponentModel.TypeConverter: [Not fixed in this PR] There is a call to [Enum.GetValues](https://github.com/dotnet/runtime/blob/main/src/libraries/System.ComponentModel.TypeConverter/src/System/ComponentModel/EnumConverter.cs#L174). This might get fixed with #72140
- Added the EnableAOTAnalyzer flag to the project to ensure that the libraries will be AOT warnings free

### CODE REVIEW

File: src/libraries/System.Reflection.DispatchProxy/src/System.Reflection.DispatchProxy.csproj [hunk 1]
```diff
@@ -2,6 +2,8 @@
   <PropertyGroup>
     <AllowUnsafeBlocks>true</AllowUnsafeBlocks>
     <TargetFramework>$(NetCoreAppCurrent)</TargetFramework>
+    <!-- This library has been annotated to be AOT safe -->
```
Comment (jkotas) [inline src/libraries/System.Reflection.DispatchProxy/src/System.Reflection.DispatchProxy.csproj:5]: Nothing in System.Reflection.DispatchProxy works with native AOT. It is good that we have it annotated, but I would not call it AOT safe.

File: src/libraries/System.Data.Common/src/System/Data/DataTable.cs [hunk 2]
```diff
@@ -3516,6 +3516,8 @@ record = NewRecord(-1);
 
         // Prevent inlining so that reflection calls are not moved to caller that may be in a different assembly that may have a different grant set.
         [MethodImpl(MethodImplOptions.NoInlining)]
+        [UnconditionalSuppressMessage("AotAnalysis", "IL3050:RequiresDynamicCode",
```
Comment (jkotas) [inline src/libraries/System.Data.Common/src/System/Data/DataTable.cs:3519]: What makes the array type available at runtime?

File: src/libraries/System.ComponentModel.TypeConverter/src/System/ComponentModel/EnumConverter.cs [hunk 3]
```diff
@@ -115,6 +115,8 @@ private static long GetEnumValue(bool isUnderlyingTypeUInt64, Enum enumVal, Cult
         /// <summary>
         /// Converts the given value object to the specified destination type.
         /// </summary>
+        [UnconditionalSuppressMessage("AotAnalysis", "IL3050:RequiresDynamicCode",
```
Comment (jkotas) [inline src/libraries/System.ComponentModel.TypeConverter/src/System/ComponentModel/EnumConverter.cs:118]: The concern is whether the array of the enum is available. I do not see anything that guarantees it.

File: src/libraries/System.Data.Common/src/System/Data/DataTable.cs (hunk 2 above)
Comment (LakshanF) [reply src/libraries/System.Data.Common/src/System/Data/DataTable.cs]: Thanks! Propagating the warnings is a larger task, punting for a later time

File: src/libraries/System.ComponentModel.TypeConverter/src/System/ComponentModel/EnumConverter.cs (hunk 3 above)
Comment (LakshanF) [reply src/libraries/System.ComponentModel.TypeConverter/src/System/ComponentModel/EnumConverter.cs:118]: Created an issue for this, #72140 and will remove this in the next commit

**Review by jkotas — APPROVED** [review]

**Pull request merged**